services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,256 @@
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+
+/**
+ * Implementation of {@link MessagingChannel} for kafka.
+ */
+@Component
+public class KafkaMessagingChannel implements MessagingChannel {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaMessagingChannel.class);
+
+    /**
+     * Requests that are in progress of processing.
+     */
+    private Map<String, CompletableFuture<InfoData>> pendingRequests = new ConcurrentHashMap<>();
+    private Map<String, CompletableFuture<List<InfoData>>> pendingChunkedRequests = new ConcurrentHashMap<>();
+
+    /**
+     * Chains of chunked messages, it is filling by messages one by one as soon as the next linked message is received.
+     */
+    private Map<String, List<ChunkedInfoMessage>> messagesChains;
+
+    /**
+     * The storage for received chunked message ids. It is needed to identify whether we have already received specific
+     * chunked message or not in order to do not have duplicates, because current version of kafka do not guarantee
+     * exactly once delivery.
+     */
+    private Map<String, Set<String>> chunkedMessageIdsPerRequest = new HashMap<>();
+
+    @Value("${northbound.messages.expiration.minutes}")
+    private int expiredTime;
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * Creates storage for chains of messages.
+     */
+    @PostConstruct
+    public void setUp() {
+        messagesChains = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoData> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+        pendingRequests.put(message.getCorrelationId(), future);
+
+        ListenableFuture<SendResult<String, Message>> futureResult = messageProducer.send(topic, message);
+        futureResult.addCallback(
+                success -> { },
+                error -> future.completeExceptionally(new MessageNotSentException(error.getMessage()))
+        );
+
+        return future.whenComplete((response, error) -> pendingRequests.remove(message.getCorrelationId()));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<InfoData>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<InfoData>> future = new CompletableFuture<>();
+        pendingChunkedRequests.put(message.getCorrelationId(), future);
+
+        ListenableFuture<SendResult<String, Message>> futureResult = messageProducer.send(topic, message);
+        futureResult.addCallback(
+                sentResult -> {
+                    messagesChains.put(message.getCorrelationId(), new ArrayList<>());
```

Comment:
To be completely safe, this operation should be done before sending request i.e. in current state there is a possible race condition - if response will be processed before processing of this callbaback.